File: app/src/main/java/com/song/sunset/activitys/ComicDetailActivity.java
package com.song.sunset.activitys;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Toast;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.ImageRequest;
import com.song.sunset.activitys.base.BaseActivity;
import com.song.sunset.beans.ComicLocalCollection;
import com.song.sunset.utils.GreenDaoUtil;
import com.song.sunset.utils.ViewUtil;
import com.song.sunset.beans.basebeans.BaseBean;
import com.song.sunset.beans.ComicDetailBean;
import com.song.sunset.R;
import com.song.sunset.adapters.ComicDetailAdapter;
import com.song.sunset.utils.loadingmanager.ProgressLayout;
import com.song.sunset.utils.rxjava.RxUtil;
import com.song.sunset.utils.retrofit.RetrofitCallback;
import com.song.sunset.utils.BitmapUtil;
import com.song.sunset.utils.retrofit.RetrofitFactory;
import com.song.sunset.utils.api.U17ComicApi;
import com.song.sunset.utils.volley.SampleVolleyFactory;
import com.sunset.greendao.gen.ComicLocalCollectionDao;

import rx.Observable;

/**
 * Created by Song on 2016/8/29 0029.
 * Email:z53520@qq.com
 */
public class ComicDetailActivity extends BaseActivity {

    private ProgressLayout progressLayout;
    public static final String COMIC_ID = "comic_id";
    private int comicId = -1;
    private ComicDetailAdapter adapter;
    private int color;
    private Toolbar toolbar;
    private FloatingActionButton floatingActionButton;
    private ComicLocalCollectionDao comicLocalCollectionDao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_comic_detail);

        comicLocalCollectionDao = GreenDaoUtil.getDaoSession().getComicLocalCollectionDao();

        progressLayout = (ProgressLayout)findViewById(R.id.progress);
        progressLayout.showLoading();

        color = Color.WHITE;
        if (getIntent() != null) {
            comicId = getIntent().getIntExtra(COMIC_ID, -1);
        }

        floatingActionButton = (FloatingActionButton) findViewById(R.id.id_comic_detail_fab);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.id_comic_detail_recycler);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        adapter = new ComicDetailAdapter(this);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        final int[] distance = {0};
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                distance[0] += dy;
                int offset = distance[0];
                float rate = offset * 1.0f / ViewUtil.dip2px(200);
                rate = rate > 1 ? 1 : rate;
                rate = rate < 0 ? 0 : rate;
                toolbar.setAlpha(rate);
            }
        });
        getDataFromRetrofit2ByObservable();
    }

//    public void getDataFromRetrofit2() {
//        Call<BaseBean<ComicDetailBean>> call = RetrofitApiBuilder.getRetrofitApi(RetrofitApi.class).queryComicDetailRDByCall(comicId);
//        RetrofitCall.call(call, new RetrofitCallback<ComicDetailBean>() {
//            @Override
//            public void onSuccess(ComicDetailBean comicDetailBean) {
//                mLoadingAndRetryManager.showContent();
//                setFloatButtonOnClick(comicDetailBean.getComic());
//                toolbar.setTitle(comicDetailBean.getComic().getName());
//                toolbar.setLogo(R.mipmap.logo);
//                adapter.setData(comicDetailBean);
//                setExtractionColorFromBitmap(comicDetailBean);
//            }
//
//            @Override
//            public void onFailure(int errorCode, String errorMsg) {
//                mLoadingAndRetryManager.showRetry();
//            }
//        });
//    }

    public void getDataFromRetrofit2ByObservable() {
        Observable<BaseBean<ComicDetailBean>> observable = RetrofitFactory.createApi(U17ComicApi.class).queryComicDetailRDByObservable(comicId);
        RxUtil.comicSubscribe(observable, new RetrofitCallback<ComicDetailBean>() {
            @Override
            public void onSuccess(ComicDetailBean comicDetailBean) {
                progressLayout.showContent();
                setFloatButtonOnClick(comicDetailBean);
                toolbar.setTitle(comicDetailBean.getComic().getName());
                toolbar.setLogo(R.mipmap.logo);
                adapter.setData(comicDetailBean);
                setExtractionColorFromBitmap(comicDetailBean);
            }

            @Override
            public void onFailure(int errorCode, String errorMsg) {
                progressLayout.showError(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        progressLayout.showLoading();
                        getDataFromRetrofit2ByObservable();
                    }
                });
            }
        });
    }

    private void setFloatButtonOnClick(final ComicDetailBean comic) {
        final long comicId = Long.parseLong(comic.getComic().getComic_id());
        if (comicLocalCollectionDao.load(comicId) == null) {
            floatingActionButton.setImageDrawable(getResources().getDrawable(android.R.drawable.star_big_off));
        } else {
            floatingActionButton.setImageDrawable(getResources().getDrawable(android.R.drawable.star_big_on));
        }
        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (comicLocalCollectionDao == null) {
                    return;
                }
                if (comicLocalCollectionDao.load(comicId) == null) {
                    comicLocalCollectionDao.insert(new ComicLocalCollection(comic.getComic().getCover(), comic.getComic().getName(),
                            comicId, comic.getComic().getDescription(), comic.getComic().getAuthor().getName(), String.valueOf(comic.getChapter_list().size())));
                    floatingActionButton.setImageDrawable(getResources().getDrawable(android.R.drawable.star_big_on));
                    Toast.makeText(getApplication(), "收藏成功", Toast.LENGTH_SHORT).show();
                } else {
                    comicLocalCollectionDao.deleteByKey(comicId);
                    floatingActionButton.setImageDrawable(getResources().getDrawable(android.R.drawable.star_big_off));
                    Toast.makeText(getApplication(), "取消收藏", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    public static void start(Context context, int comicId) {
        Intent intent = new Intent(context, ComicDetailActivity.class);
        intent.putExtra(COMIC_ID, comicId);
        context.startActivity(intent);
    }

    public void setExtractionColorFromBitmap(ComicDetailBean comicDetailRD) {
        RequestQueue queue = SampleVolleyFactory.getRequestQueue(this);
        ImageRequest imageRequest = new ImageRequest(comicDetailRD.getComic().getCover(),
                new Response.Listener<Bitmap>() {
                    @Override
                    public void onResponse(Bitmap response) {
                        color = BitmapUtil.getColorFromBitmap(response);
                        adapter.setColor(color);
                        toolbar.setBackgroundColor(color);
                    }
                }, 128, 128, Bitmap.Config.ALPHA_8, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
            }
        });
        imageRequest.setRetryPolicy(new DefaultRetryPolicy());
        queue.add(imageRequest);
    }
}

File: app/src/main/java/com/song/sunset/activitys/ComicDetailActivity.kt
package com.song.sunset.activitys

import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.android.arouter.facade.annotation.Route
import com.facebook.common.references.CloseableReference
import com.facebook.datasource.DataSource
import com.facebook.imagepipeline.datasource.BaseBitmapDataSubscriber
import com.facebook.imagepipeline.image.CloseableImage
import com.song.core.statusbar.StatusBarUtil
import com.song.sunset.R
import com.song.sunset.adapters.ComicDetailAdapter
import com.song.sunset.adapters.ComicDetailAdapter.COMIC_LIST_TYPE
import com.song.sunset.beans.ComicDetailBean
import com.song.sunset.utils.BitmapUtil
import com.song.sunset.utils.ViewUtil
import com.song.sunset.utils.fresco.FrescoUtil
import com.song.sunset.viewmodel.ComicDetailViewModel
import kotlinx.android.synthetic.main.activity_comic_detail_mvvm.*

@Route(path = "/song/comic/detail")
open class ComicDetailActivity : AppCompatActivity() {

    companion object {
        const val COMIC_ID = "comic_id"
    }

    var color = Color.WHITE
    private var comicId = -1
    private var adapter: ComicDetailAdapter? = null
    private var comicDetailBean: ComicDetailBean? = null
    private lateinit var vm: ComicDetailViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //设置状态栏透明
        StatusBarUtil.setTransparent(this)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT

        setContentView(R.layout.activity_comic_detail_mvvm)

        initView()

        addListener()

        addObserver()

        vm.getComicDetailData(comicId)
        vm.getCollectionStatus(comicId)
    }

    private fun addObserver() {
        vm.comicDetailBean.observe(this, Observer {
            comicDetailBean = it
            toolbar.title = it.comic.name
            adapter?.setData(it)
            setExtractionColorFromBitmap(it)
            vm.updateCollectedComicData(it)
        })

        vm.dataStatus.observe(this, Observer {
            if (it) {
                progress.showContent()
                id_comic_detail_fab.visibility = View.VISIBLE
            } else {
                progress.showError {
                    progress.showLoading()
                    id_comic_detail_fab.visibility = View.GONE
                    vm.getComicDetailData(comicId)
                }
            }
        })

        vm.collectionStatus.observe(this, Observer {
            val res = if (it) android.R.drawable.star_big_on else android.R.drawable.star_big_off
            id_comic_detail_fab.setImageDrawable(resources.getDrawable(res))
        })

        vm.showTips.observe(this, Observer {
            Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
        })
    }

    private fun addListener() {
        var distance = 0
        id_comic_detail_recycler.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                distance += dy
                var rate = distance * 1.0f / ViewUtil.dip2px(200F)
                if (rate > 1) {
                    rate = 1F
                } else if (rate < 0) {
                    rate = 0F
                }
                toolbar.alpha = rate
            }
        })

        id_comic_detail_fab.setOnClickListener {
            vm.changeCollectedStatus(comicDetailBean!!)
        }
    }

    private fun initView() {
        vm = ViewModelProviders.of(this).get(ComicDetailViewModel::class.java)
        progress.showLoading()

        if (intent != null) {
            comicId = intent.getIntExtra(COMIC_ID, -1)
        }
        adapter = ComicDetailAdapter(this)
        id_comic_detail_recycler.adapter = adapter
        val layoutManager = GridLayoutManager(this, 3)
        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val itemViewType = adapter?.getItemViewType(position)
                return if (itemViewType == COMIC_LIST_TYPE) {
                    1
                } else {
                    3
                }
            }
        }
        id_comic_detail_recycler.layoutManager = layoutManager
    }

    private fun setExtractionColorFromBitmap(comicDetailRD: ComicDetailBean) {
        FrescoUtil.getCachedImageBitmap(FrescoUtil.getDataSource(comicDetailRD.comic.cover)!!, object : BaseBitmapDataSubscriber() {
            override fun onNewResultImpl(bitmap: Bitmap?) {
                if (this@ComicDetailActivity.isFinishing) return
                color = BitmapUtil.getColorFromBitmap(bitmap)
                Handler().post { adapter?.setColor(color) }
                toolbar.setBackgroundColor(color)
            }

            override fun onFailureImpl(dataSource: DataSource<CloseableReference<CloseableImage>>) {

            }
        })
    }
}
